xds/src/main/java/io/grpc/xds/XdsLrsClient.java

```diff
@@ -0,0 +1,358 @@
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Client of XDS load reporting service. Methods in this class are expected to be called in
+ * the same synchronized context that {@link XdsLoadBalancer#helper#getSynchronizationContext}
+ * returns.
+ */
+@NotThreadSafe
+final class XdsLrsClient implements XdsLoadStatsManager {
+
+  @VisibleForTesting
+  static final String TRAFFICDIRECTOR_HOSTNAME_FIELD
+      = "com.googleapis.trafficdirector.grpc_hostname";
+  private final String serviceName;
+  private final ManagedChannel channel;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timerService;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final Stopwatch retryStopwatch;
+  private final ChannelLogger logger;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final XdsLoadReportStore loadReportStore;
+
+  @Nullable
+  private BackoffPolicy lrsRpcRetryPolicy;
+  @Nullable
+  private ScheduledHandle lrsRpcRetryTimer;
+
+  @Nullable
+  private LrsStream lrsStream;
+
+  XdsLrsClient(ManagedChannel channel,
+      Helper helper,
+      BackoffPolicy.Provider backoffPolicyProvider) {
+    this.channel = checkNotNull(channel, "channel");
+    this.serviceName = checkNotNull(helper.getAuthority(), "serviceName");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.stopwatchSupplier = GrpcUtil.STOPWATCH_SUPPLIER;
+    this.retryStopwatch = stopwatchSupplier.get();
+    this.logger = checkNotNull(helper.getChannelLogger(), "logger");
+    this.timerService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.loadReportStore = new XdsLoadReportStore(serviceName);
+  }
+
+  @VisibleForTesting
+  XdsLrsClient(ManagedChannel channel,
+      Helper helper,
+      Supplier<Stopwatch> stopwatchSupplier,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      XdsLoadReportStore loadReportStore) {
+    this.channel = checkNotNull(channel, "channel");
+    this.serviceName = checkNotNull(helper.getAuthority(), "serviceName");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.retryStopwatch = stopwatchSupplier.get();
+    this.logger = checkNotNull(helper.getChannelLogger(), "logger");
+    this.timerService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.loadReportStore = checkNotNull(loadReportStore, "loadReportStore");
+  }
+
+  @Override
+  public void startLoadReporting() {
+    checkState(lrsStream == null, "previous lbStream has not been cleared yet");
```

Comment:
I think this method, given its content, should be a private method. It being a public method has two issues:
1. `lrsStream` can be altered by internal logic of this class, which is beyond caller's control. A precondition should be something that the caller can control, e.g., if a method says it can only be called once, the caller can conform to that.
2. It's called from both internal and external. As this is an interface method which will be mocked when testing the caller (XdsLoadBalanacer), the internal call path will be out of the picture, making the test impossible to simulate what's happening on the interface in reality.

My suggestion is that you change this to a `private void startLrStream()`, and make the public `startLoadReporting()` call it. The internal caller should call the private method. Since `XdsLoadBalancer` is expected to call it at most once, you can add a `boolean started` and have the preconditions in `startLoadReporting()` to be `checkState(!started)`.

Reply:
ACK.